Sentry를 통해 구체적인 에러 상황을 탐지하도록 개선

Cllaude99 opened the pull request:
Resolves: #237 

## PR 유형
- [x] 새로운 기능 추가

## 작업 내용

### 🚨 [문제 상황]
이전에는 애플리케이션에서 발생하는 예외와 에러 상황을 효과적으로 모니터링하고 분석할 수 있는 시스템이 없었습니다. 이러한 이유로 사용자 경험 저하와 직결되는 오류들을 실시간으로 파악하고 해결하기 어려웠으며, 프로덕션 환경에서 발생하는 문제들을 재현하기 위한 상세 정보가 부족했습니다.

### 🛠️ [문제 해결 과정]
위의 이유로 Sentry를 도입하여 애플리케이션의 에러 모니터링 시스템을 구축했습니다. @sentry/react 패키지를 설치하고 React 애플리케이션에 적합한 초기화 설정을 구현했습니다. 개발 및 프로덕션 환경에 맞는 설정을 분리하고, 브라우저 트레이싱과 세션 리플레이 기능을 통합했습니다. 

`sentry.ts`

``` javascript
import * as Sentry from '@sentry/react';

export const initSentry = () => {
  Sentry.init({
    // 여기에 실제 Sentry DSN을 입력하세요
    dsn: import.meta.env.VITE_SENTRY_DSN,

    // 개발 환경에서만 에러 보고 활성화 (선택적)
    enabled: import.meta.env.PROD,

    // 기본 PII 정보 전송 활성화 (사용자 식별을 위한 헤더 및 IP)
    sendDefaultPii: true,

    integrations: [
      // 브라우저 트레이싱 통합
      Sentry.browserTracingIntegration(),

      // 세션 리플레이 통합 (선택적)
      Sentry.replayIntegration(),
    ],

    // 트레이스 샘플링 비율 설정 (0.0 ~ 1.0)
    tracesSampleRate: 0.5,

    // 요청 추적 대상 URL 패턴
    tracePropagationTargets: [
      /^\//,
      /^https:\/\/www\.api\.syncspot\.kr/,
      /^https:\/\/www\.dev\.syncspot\.kr/,
    ],

    // 세션 리플레이 설정 (선택적)
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,

    // 환경 설정 (개발/테스트/프로덕션)
    environment: import.meta.env.MODE,
  });
};

export const sentryReactErrorHandler = Sentry.reactErrorHandler;

```

해당사항과는 별개로 작업내용에 대해 PR을 보내는 과정에서 CI/CD 환경에서의 패키지 설치 문제가 발생하였습니다. 이를 해결하기 위해 Yarn의 hardened 모드를 비활성화하는 설정을 추가했습니다.

### 🎯 [결과]
이제 애플리케이션에서 발생하는 모든 종류의 에러(렌더링 에러, API 요청 에러, 비동기 처리 에러 등)를 실시간으로 모니터링하고 분석할 수 있게 되었습니다. 에러 발생 시 사용자 컨텍스트, 브라우저 정보, 세션 데이터 등이 함께 수집되어 문제 해결에 필요한 모든 정보를 확보할 수 있습니다. 


## 스크린샷
<img width="584" alt="스크린샷 2025-05-06 오후 4 12 15" src="https://github.com/user-attachments/assets/404afff7-8b2d-4b8d-bb5c-4b30507ba149" />
<img width="958" alt="스크린샷 2025-05-06 오후 4 12 41" src="https://github.com/user-attachments/assets/56069e06-9bfa-4673-a8b4-275daedfdbd6" />


테스트를 해본 결과 Sentry 대시보드에서 아래와 같이 테스트에 대한 이벤트를 캡쳐할 수 있었습니다.
첨부된 이미지를 보면 "Proxy test message from SentryTest component" 메시지가 성공적으로 캡처된 것을 확인할 수 있습니다.

## PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신규 기능**
	- Sentry 오류 추적 및 모니터링 기능이 앱에 통합되었습니다.
	- 사용자 동의에 따라 오류 추적 활성화/비활성화가 가능합니다.
	- 오류 발생 시 Sentry 기반 에러 바운더리 및 라우터 통합이 적용되었습니다.
	- Sentry를 통한 사용자 정보, 컨텍스트, 태그 추가 기능이 제공됩니다.

- **버그 수정**
	- 오류 분류 및 알림 메시지 처리가 개선되었습니다.

- **환경설정**
	- Sentry 관련 패키지 및 의존성이 추가되었습니다.
	- Yarn 설정이 일부 업데이트되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Cllaude99 committed:
chore: Sentry 패키지 버전 9.15.0 추가 및 관련 의존성 업데이트

- @sentry/react, @sentry/browser, @sentry/core 및 내부 Sentry 패키지 추가
- package.json 및 yarn.lock 파일에 Sentry 관련 의존성 추가
- .pnp.cjs 파일에 Sentry 패키지 정보 업데이트

Cllaude99 committed:
feat: Sentry 에러 캡처 및 처리 로직 추가

- Axios 에러와 일반 자바스크립트 에러를 구분하여 처리하는 로직 추가
- Sentry를 통해 API 에러 및 일반 에러를 기록하도록 구현
- 사용자에게 오류 알림을 표시하는 기능 개선
- Sentry 초기화 코드 추가

Cllaude99 committed:
fix: 리프레쉬 토큰 처리 로직 개선

- catch 블록에서 에러를 무시하도록 수정하여 코드 간결화
- 리프레쉬 토큰 만료 시 로컬 스토리지 클리어 및 리다이렉트 처리 유지

Cllaude99 committed:
feat: SentryErrorBoundary 컴포넌트 추가 및 라우터 통합

- SentryErrorBoundary 컴포넌트를 새로 추가하여 Sentry와의 에러 처리를 통합
- AppRouter에서 UnknownErrorBoundary를 SentryErrorBoundary로 변경하여 Sentry 에러 캡처 기능 적용
- APIErrorBoundary에 SentryErrorBoundary를 래핑하여 에러 처리 로직 개선

Cllaude99 committed:
feat: Sentry 통합을 위한 유틸리티 파일 추가

- Sentry 초기화 및 에러 캡처를 위한 유틸리티 함수 추가
- 사용자 정보 및 컨텍스트 설정을 위한 함수 구현
- React Router와의 통합을 위한 래퍼 함수 추가
- API 에러 캡처를 위한 함수 추가

Cllaude99 committed:
chore: disable Yarn hardened mode to fix package installation issues

Cllaude99 committed:
chore: .pnp.cjs 및 yarn.lock 파일 업데이트

- .pnp.cjs 파일에서 Sentry 패키지 버전 9.15.0 추가 및 관련 의존성 정보 수정
- yarn.lock 파일에 Sentry 및 Storybook 관련 패키지 버전 업데이트

Cllaude99 committed:
chore: code-rule.mdc 파일 삭제

- 프론트엔드 디자인 가이드라인 문서인 code-rule.mdc 파일을 삭제하여 코드베이스 정리

Cllaude99 committed:
chore: Sentry 초기화 코드 정리 및 주석 수정

- Sentry 초기화 함수에서 불필요한 주석 제거 및 코드 정리
- 환경 변수 및 설정 관련 주석을 명확하게 수정하여 가독성 향상

Cllaude99 commented:
@chae-dahee 코드 리뷰 부탁드립니다.

chae-dahee commented on src/App.tsx:
```diff
@@ -13,13 +16,39 @@ const queryClient = new QueryClient({
+          });
+        } else {
+          // 기타 알 수 없는 에러 타입
+          captureException('Unknown error occurred');
```
여기에서 문자열로 전달하지 말고, Error 객체로 감싸서 전달하면 captureException 의 string 타입 및 로직을 정리할 수 있을 것 같습니다.
captureException(new Error('Unknown error occurred'));

Cllaude99 replied:
네 반영했습니다.

chae-dahee commented on src/App.tsx:
```diff
@@ -13,13 +16,39 @@ const queryClient = new QueryClient({
+          // Sentry에 API 에러 기록
+          captureApiError(error.config?.url || 'unknown', error);
+
+          // 사용자에게 오류 알림 표시
+          CustomToast({
+            type: TOAST_TYPE.ERROR,
+            status: errorData.status,
+            message: errorData.message,
+          });
```
반복되는 customToast 생성을 공통 토스트메시지로 묶어서 처리할 수 있을 것 같습니다~

Cllaude99 replied:
> 반복되는 customToast 생성을 공통 토스트메시지로 묶어서 처리할 수 있을 것 같습니다~

혹시 이게 어떤말씀이실까요? CustomToast 자체로도 공통적으로 처리하고 있다고 판단했어요.

chae-dahee replied:
```
  const message =
    isAxios && isAxiosError(error)
      ? getErrorData(error).message
      : error instanceof Error
        ? error.message
        : '알 수 없는 오류가 발생했습니다.';

  CustomToast({
    type: TOAST_TYPE.ERROR,
    status: 'error',
    message,
  });
```
if 문에 따른 반복적인 CustomToast 코드 작성을 줄이는 방식을 생각했습니다! message만 다르므로 이부분만 처리해주면 될 것 같아요

Cllaude99 replied:
반영했습니다~

chae-dahee commented on src/components/error/boundary/SentryErrorBoundary.tsx:
App.tsx의 onError 에서 Sentry에 에러 기록 및 전달하면서 Sentry.captureException() 이 호출됩니다.
이후 throw 된 에러가 SentryErrorBoundary 에서 포착되어 내부적으로 한번더 captureException() 이 호출되어 중복 전송될 가능성이 있는 것 같습니다.
혹시 위 App.tsx 의 onError 설정에 대하여 참고한 문서가 있을까요? 또는 설정하신 동작방식을 설명해주실 수 있나요?

Cllaude99 replied:
현재 구조에서는 App.tsx의 onError에서 Sentry에 에러를 보고한 후, 같은 에러가 SentryErrorBoundary까지 전파되면 중복으로 보고되어 말씀하신대로 중복이 발생할 것 같습니다.

따라서 이를 해결하기 위해 에러 전파를 차단하는 대신 App.tsx에서 에러를 캡처할 때 _sentryReported 플래그를 추가하여 이미 Sentry에 보고된 에러임을 표시하도록 하였습니다. 또 SentryErrorBoundary에 beforeCapture 훅을 추가하여, 이미 _sentryReported 플래그가 설정된 에러는 중복 보고되지 않도록 scope.clear()를 호출하도록 하였습니다.
위와 같이 처리하여 Sentry에 중복 보고되지 않도록 수정하였습니다.  

onError설정과 관련해서는 공식문서의 내용을 참고하였습니다. [Sentry 공식 문서](https://docs.sentry.io/platforms/javascript/guides/react/components/errorboundary/)를 참고하시면 도움이 될 것 같습니다.

chae-dahee commented on src/utils/sentryUser.ts:
```diff
@@ -0,0 +1,56 @@
+    // 사용자가 로그인했을 때
+    Sentry.setUser({
+      id: user.id,
+      email: user.email,
```
이메일의 경우, 사용자의 개인정보입니다.
또한 src/utils/sentry.ts 의 sendDefaultPii 또한 사용자의 여러 정보를 담고있다고 알고있습니다.
혹시 위 정보가 어떤 부분에서 필요한가요? 이 경우, 개인정보 동의 선택지를 통해 사용자에게 고지를 해야하는건 아닌지 궁굼합니다.

Cllaude99 replied:
네, 개인정보 보호에 대한 중요한 지적 감사합니다. 
검토 결과 실제로 현재 코드는 setSentryUser 함수에서 이메일 등 개인식별정보를 Sentry에 전송하고, sendDefaultPii: true 설정으로 IP 주소와 쿠키 등 추가 개인정보도 자동 수집하고 있었습니다. 
이 문제를 해결하기 위해 sentry.ts에서 sendDefaultPii: false로 변경하여 기본 개인식별정보 자동 수집을 비활성화하고, 사용자 동의 여부에 따라 Sentry 활성화를 제어하는 로직을 추가했습니다. 또한 sentryUser.ts에서는 사용자 동의 여부에 따라 이메일 등 개인정보를 포함할지 결정하는 파라미터를 추가하고, 동의하지 않은 경우에는 익명화된 ID만 전송하도록 개선했습니다.

chae-dahee replied:
좋습니다 👍 
개인정보 동의 여부는 사용자에게 어떤 방식으로 표시되나요?

Cllaude99 replied:
> 좋습니다 👍 개인정보 동의 여부는 사용자에게 어떤 방식으로 표시되나요?

개인 정보 동의 여부는 개인정보를 동의한다는 모달을 띄우는 방식으로 해서 사용자의 동의를 받아야할 것 같아요.
이 부분은 추가 구현이 필요할 것 같아서 일단 기본적으로는 사용자의 정보 보호를 위해 개인정보를 수집하지 않고 있고 추후에 동의 모달에서 응답값을 받아와 수정할 수 있을 것 같습니다.

chae-dahee reviewed (commented):
에러 모니터링 도구로 Sentry 도입하시느라 고생하셨습니다~
Sentry 무료사용은 에러 이슈 조회할 수 있는 계정을 하나로 제한하고있는 걸로 알고있습니다. 싱크스팟 구글계정을 사용하는건 어떠신가요?
또한 에러 발생 시의 실제 코드를 참고할 수 있도록 soucemap 을 sentry에 제공하는 방식을 추가하는것은 어떤가요? ([참고글](https://blog.runners.im/sentry-%EC%97%90%EB%9F%AC-%EB%A1%9C%EA%B9%85-%EA%B0%80%EC%9D%B4%EB%93%9C-with-sourcemaps-fd27cbc9dd00))

Cllaude99 committed:
fix: 에러 캡처 로직 개선

- Sentry의 에러 캡처를 위한 코드 수정
- 'Unknown error occurred' 메시지를 Error 객체로 감싸서 보다 명확한 에러 정보를 제공하도록 변경

Cllaude99 committed:
feat: Sentry 에러 보고 플래그 추가

- Sentry 에러 인터페이스에 보고 플래그 추가
- 에러 발생 시 중복 보고 방지를 위한 로직 구현
- SentryErrorBoundary에서 이미 보고된 에러 확인 기능 추가

Cllaude99 committed:
feat: Sentry 사용자 동의 및 개인정보 처리 로직 추가

- 사용자 에러 추적 동의 여부를 확인하여 Sentry 활성화 조건 추가
- 개인정보 수집 동의에 따라 Sentry 사용자 정보 설정 로직 개선
- 사용자 ID 익명화를 위한 해시 함수 추가

Cllaude99 committed:
fix: 에러 처리 로직 개선 및 사용자 알림 통합

- Axios 에러와 일반 에러에 대한 처리 로직을 통합하여 코드 간결화
- 사용자에게 오류 알림을 표시하는 로직을 개선하여 중복 코드를 제거하고 가독성 향상
- 에러 메시지와 상태를 결정하는 로직을 추가하여 사용자 경험 개선
